File: pathfinder-core/src/main/java/de/cubbossa/pathfinder/node/selection/attribute/GroupSelectionAttribute.java
package de.cubbossa.pathfinder.node.selection.attribute;

import com.mojang.brigadier.arguments.ArgumentType;
import de.cubbossa.pathfinder.PathFinderProvider;
import de.cubbossa.pathfinder.group.NodeGroup;
import de.cubbossa.pathfinder.misc.NamespacedKey;
import de.cubbossa.pathfinder.node.Node;
import de.cubbossa.pathfinder.node.selection.AbstractNodeSelectionParser;
import de.cubbossa.pathfinder.node.selection.NodeSelectionAttribute;
import de.cubbossa.pathfinder.storage.StorageUtil;
import de.cubbossa.pathfinder.util.SelectionParser;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import lombok.Getter;
import org.pf4j.Extension;

@Getter
@Extension(points = NodeSelectionAttribute.class)
public class GroupSelectionAttribute implements NodeSelectionAttribute<Collection<NodeGroup>> {

  private final String key = "group";

  @Override
  public ArgumentType<Collection<NodeGroup>> getValueType() {
    return r -> {
      String in = r.getRemaining();
      Collection<NodeGroup> groups = new HashSet<>();
      NamespacedKey key;
      try {
        key = NamespacedKey.fromString(in);
      } catch (Throwable t) {
        throw new IllegalArgumentException("Invalid namespaced key: '" + in + "'.");
      }
      Optional<NodeGroup> group = PathFinderProvider.get().getStorage().loadGroup(key).join();
      groups.add(group.orElseThrow(
          () -> new IllegalArgumentException("There is no group with the key '" + key + "'")));
      return groups;
    };
  }

  @Override
  public Type getAttributeType() {
    return Type.FILTER;
  }

  @Override
  public List<Node> execute(AbstractNodeSelectionParser.NodeArgumentContext<Collection<NodeGroup>> context) {
    return context.getScope().stream()
        .filter(node -> StorageUtil.getGroups(node).containsAll(context.getValue()))
        .collect(Collectors.toList());
  }

  @Override
  public List<String> getStringSuggestions(SelectionParser.SuggestionContext context) {
    return PathFinderProvider.get().getStorage().loadAllGroups().join().stream()
        .map(NodeGroup::getKey)
        .map(NamespacedKey::toString)
        .collect(Collectors.toList());
  }
}

File: pathfinder-core/src/main/java/de/cubbossa/pathfinder/node/selection/attribute/GroupSelectionAttribute.kt
package de.cubbossa.pathfinder.node.selection.attribute

import com.mojang.brigadier.arguments.ArgumentType
import de.cubbossa.pathfinder.PathFinder
import de.cubbossa.pathfinder.group.NodeGroup
import de.cubbossa.pathfinder.misc.NamespacedKey
import de.cubbossa.pathfinder.misc.NamespacedKey.Companion.fromString
import de.cubbossa.pathfinder.node.Node
import de.cubbossa.pathfinder.node.selection.AbstractNodeSelectionParser.NodeArgumentContext
import de.cubbossa.pathfinder.node.selection.NodeSelectionAttribute
import de.cubbossa.pathfinder.storage.StorageUtil
import de.cubbossa.pathfinder.util.SelectionParser
import kotlinx.coroutines.runBlocking
import lombok.Getter
import org.pf4j.Extension
import java.util.stream.Collectors

@Getter
@Extension(points = [NodeSelectionAttribute::class])
class GroupSelectionAttribute : NodeSelectionAttribute<Collection<NodeGroup>> {

    override val key = "group"

    override val valueType: ArgumentType<Collection<NodeGroup>>
        get() {
            return ArgumentType<Collection<NodeGroup>> {
                runBlocking {
                    val input = it.remaining
                    val groups: MutableCollection<NodeGroup> = HashSet()
                    val key: NamespacedKey
                    try {
                        key = fromString(input)
                    } catch (t: Throwable) {
                        throw IllegalArgumentException("Invalid namespaced key: '$input'.")
                    }
                    val group = PathFinder.get().storage.loadGroup(key)
                        ?: throw IllegalArgumentException("There is no group with the key '$key'")
                    groups.add(group)
                    groups
                }
            }
        }

    override val attributeType: NodeSelectionAttribute.Type
        get() = NodeSelectionAttribute.Type.FILTER

    override fun execute(context: NodeArgumentContext<Collection<NodeGroup>>): List<Node> {
        return context.scope.stream()
            .filter { StorageUtil.getGroups(it).containsAll(context.value) }
            .collect(Collectors.toList())
    }

    override fun getStringSuggestions(context: SelectionParser.SuggestionContext): List<String> =
        runBlocking {
            PathFinder.get().storage.loadAllGroups().stream()
                .map(NodeGroup::key)
                .map(NamespacedKey::toString)
                .collect(Collectors.toList())

        }
}
